server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java

```diff
@@ -781,12 +781,14 @@ void allocateUnassigned() {
          */
         ShardRouting[] unassignedShards = unassigned.drain();
         List<ShardRouting> allUnassignedShards = Arrays.stream(unassignedShards).collect(Collectors.toList());
-        List<ShardRouting> localUnassignedShards = allUnassignedShards.stream()
-            .filter(shard -> RoutingPool.LOCAL_ONLY.equals(RoutingPool.getShardPool(shard, allocation)))
+        List<ShardRouting> nonLocalUnassignedShards = allUnassignedShards.stream()
+            .filter(shard -> !RoutingPool.LOCAL_ONLY.equals(RoutingPool.getShardPool(shard, allocation)))
             .collect(Collectors.toList());
-        allUnassignedShards.removeAll(localUnassignedShards);
-        allUnassignedShards.forEach(shard -> routingNodes.unassigned().add(shard));
-        unassignedShards = localUnassignedShards.toArray(new ShardRouting[0]);
+
+        nonLocalUnassignedShards.forEach(shard -> routingNodes.unassigned().add(shard));
+        unassignedShards = allUnassignedShards.stream()
+            .filter(shard -> RoutingPool.LOCAL_ONLY.equals(RoutingPool.getShardPool(shard, allocation)))
+            .toArray(ShardRouting[]::new);
         ShardRouting[] primary = unassignedShards;
```

Comment:
Good find @RS146BIJAY! As @Bukhtawar and @vikasvb90 have mentioned, I think there are a few more optimizations to try here. I think maybe it could be simplified to something like the following to avoid intermediate lists and extra iterations:

```java
List<ShardRouting> primaryList = new ArrayList<>();
for (ShardRouting shard : unassigned.drain()) {
    if (RoutingPool.LOCAL_ONLY.equals(RoutingPool.getShardPool(shard, allocation))) {
        primaryList.add(shard);
    } else {
        routingNodes.unassigned().add(shard);
    }
}
ShardRouting[] primary = primaryList.toArray(new ShardRouting[0]);
```

Is there an opportunity to add a [microbenchmark](https://github.com/opensearch-project/OpenSearch/tree/8cecd5aafc61cb20f3a3ddd7e71f2fd6347f2660/benchmarks/src/main/java/org/opensearch/benchmark/routing/allocation) that can quickly show improvements or regressions?

Reply:
Remove unnecessary streams from the code. For micro benchmarking, I am using [microbenchmark suite provided by OpenSearch](https://github.com/opensearch-project/OpenSearch/tree/main/benchmarks). Major bottleneck is still List.removeAll call. Only observed 2 seconds improvements by removing unnecessary streams.

Reply:
Lets share the benchmark results and check in any additional code modified in `AllocationBenchmark`